backvendor/hg_test.go

```diff
@@ -0,0 +1,288 @@
+			Source: &GoSource{},
+			VCS:    vcs.ByCmd(vcsHg),
+		},
+	}
+
+	revision := "d4c3dbfa77a74ae238e401d5d2197b45f30d8513"
+	mockedStdout = strings.TrimSpace(`
+		<?xml version="1.0"?>
+		<log>
+		<logentry revision="1" node="d4c3dbfa77a74ae238e401d5d2197b45f30d8513">
+		<tag>tip</tag>
+		<author email="example@example.com">Example</author>
+		<date>2018-09-20T12:00:00+00:00</date>
+		<msg xml:space="preserve">example</msg>
+		</logentry>
+		</log>
+	`) + "\n"
+	tm, err := wt.TimeFromRevision(revision)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	var expected time.Time
+	expected.UnmarshalText([]byte("2018-09-20T12:00:00+00:00"))
+	if !tm.Equal(expected) {
+		t.Errorf("unexpected time: got %s, want %s", tm, expected)
+	}
+}
+
+func TestHgReachableTag(t *testing.T) {
+	defer mockExecCommand()()
+
+	wt := hgWorkingTree{
+		anyWorkingTree: anyWorkingTree{
+			Source: &GoSource{},
+			VCS:    vcs.ByCmd(vcsHg),
+		},
+	}
+
+	type tcase struct {
+		name       string
+		stdout     string
+		expSuccess bool
+		expTag     string
+	}
+	tcases := []tcase{
+		tcase{
+			name: "no-tags",
+			stdout: `
+	<?xml version="1.0"?>
+	<log>
+	</log>
+			`,
+			expSuccess: false,
+		},
+
+		tcase{
+			name: "no-semver",
+			stdout: `
+	<?xml version="1.0"?>
+	<log>
+	<logentry revision="1" node="67639326a575e4f7db1f3f70697bf096af1cbe8d">
+	<tag>v1.0.1beta1</tag>
+	<author email="example@example.com">Example</author>
+	<date>2018-09-20T12:00:00+00:00</date>
+	<msg xml:space="preserve">example</msg>
+	</logentry>
+	</log>
+	`,
+			expSuccess: true,
+			expTag:     "v1.0.1beta1",
+		},
+
+		tcase{
+			name: "semver-after-non-semver",
+			stdout: `
+	<?xml version="1.0"?>
+	<log>
+	<logentry revision="1" node="67639326a575e4f7db1f3f70697bf096af1cbe8d">
+	<tag>v1.0.1beta1</tag>
+	<author email="example@example.com">Example</author>
+	<date>2018-09-20T12:00:00+00:00</date>
+	<msg xml:space="preserve">example</msg>
+	</logentry>
+	<logentry revision="1" node="67639326a575e4f7db1f3f70697bf096af1cbe8d">
+	<tag>v1.0.0</tag>
+	<author email="example@example.com">Example</author>
+	<date>2018-09-20T12:00:00+00:00</date>
+	<msg xml:space="preserve">example</msg>
+	</logentry>
+	</log>
+	`,
+			expSuccess: true,
+			expTag:     "v1.0.0",
+		},
+	}
+
+	revision := "d4c3dbfa77a74ae238e401d5d2197b45f30d8513"
+	for _, tc := range tcases {
+		mockedStdout = strings.TrimSpace(tc.stdout) + "\n"
+		tag, err := wt.ReachableTag(revision)
+		if tc.expSuccess {
+			if err != nil {
+				t.Errorf("unexpected failure: %s: %s", tc.name, err)
+			} else if tag != tc.expTag {
+				t.Errorf("incorrect tag: %s: %s", tc.name, tag)
+			}
+		} else if err != ErrorVersionNotFound {
+			t.Errorf("unexpected error: %s: %s", tc.name, err)
+		}
+	}
+}
+
+func TestHgErrors(t *testing.T) {
+	defer mockExecCommand()()
+
+	wt := hgWorkingTree{
+		anyWorkingTree: anyWorkingTree{
+			Source: &GoSource{},
+			VCS:    vcs.ByCmd(vcsHg),
+		},
+	}
+
+	mockedStderr = "fatal: not a git repository\n"
```

Comment:
This should likely be some mercurial specific.

Reply:
Fixed.